test/config-luatest/config_info_hierarchy_test.lua

```diff
@@ -0,0 +1,153 @@
+local fio = require('fio')
+local t = require('luatest')
+local cbuilder = require('luatest.cbuilder')
+local cluster = require('test.config-luatest.cluster')
+local server = require('luatest.server')
+local it = require('test.interactive_tarantool')
+
+local g = t.group()
+
+g.before_all(cluster.init)
+g.after_each(cluster.drop)
+g.after_all(cluster.clean)
+
+g.before_test('test_script', function(g)
+    g.server = server:new({alias = 'classic_server'})
+    g.server:start()
+end)
+
+g.after_test('test_script', function(g)
+    g.server:stop()
+    g.server = nil
+end)
+
+g.before_test('test_interactive', function(g)
+    g.it = it.new()
+end)
+
+g.after_test('test_interactive', function(g)
+    g.it:close()
+    g.it = nil
+end)
```

Comment:
Nit: As I see in other tests (e.g in `enterprise-luatest`), such `before_/after_` calls surround the test on which they're applied.

```lua
g.before_test('a')
g.a = function()
g.after_test('b')
```

IMO it's a lot simpler to read the tests like this.

Reply:
The idea behind splitting these hooks from the main testing code is that setup/teardown code in not very interesting, when looking into a test case logic. For example, we have a preamble that defines a standard setup/teardown hooks for `cluster`:

```lua
g.before_all(cluster.init)                                                                                       
g.after_each(cluster.drop)                                                                                       
g.after_all(cluster.clean)  
```

And we just forget about all this machinery, when writing/looking a test case. I think that's nice.

OTOH, I understand that if you see tightly coupled things splitted over the file, it distracts a bit.

I don't like a direct solution: placing the hooks in-between the test cases like these are objects of the same kind.

However, I can write it in a more structured way:

```lua
-- Add a test case with before/after_test hooks in a convenient
-- way.
--
-- g:add_test_case('test_foo', {
--     before = function(g)
--         <...>
--     end,
--     function(g)
--     end,
--     after = function(g)
--         <...>
--     end,
-- })
g.add_test_case = function(g, name, funcs)
    if funcs.before ~= nil then
        g.before_test(name, funcs.before)
    end
    assert(funcs[1] ~= nil)
    g[name] = funcs[1]
    if funcs.after ~= nil then
        g.after_test(name, funcs.after)
    end
end

<...>

-- Verify that there are no instance/replicaset/group names in
-- config:info() if tarantool is started from a Lua script,
-- without the declarative configuration.
g:add_test_case('test_script', {
    before = function(g)
        g.server = server:new({alias = 'classic_server'})
        g.server:start()
    end,
    function(g)
        g.server:exec(function()
            local config = require('config')

            t.assert_equals(config:info().hierarchy, {})
            t.assert_equals(config:info('v1').hierarchy, {})
            t.assert_equals(config:info('v2').hierarchy, {})
            t.assert_equals(box.info.config.hierarchy, {})
        end)
    end,
    after = function(g)
        g.server:stop()
        g.server = nil
    end,
})
```

I like it more.

OTOH, I can just define one `after_each` and start needed processes from within a test case:

```lua
diff --git a/test/config-luatest/config_info_hierarchy_test.lua b/test/config-luatest/config_info_hierarchy_test.lua
index 6cd333ac25..562c47c858 100644
--- a/test/config-luatest/config_info_hierarchy_test.lua
+++ b/test/config-luatest/config_info_hierarchy_test.lua
@@ -11,23 +11,15 @@ g.before_all(cluster.init)
 g.after_each(cluster.drop)
 g.after_all(cluster.clean)
 
-g.before_test('test_script', function(g)
-    g.server = server:new({alias = 'classic_server'})
-    g.server:start()
-end)
-
-g.after_test('test_script', function(g)
-    g.server:stop()
-    g.server = nil
-end)
-
-g.before_test('test_interactive', function(g)
-    g.it = it.new()
-end)
-
-g.after_test('test_interactive', function(g)
-    g.it:close()
-    g.it = nil
+g.after_each(function(g)
+    if g.server ~= nil then
+        g.server:stop()
+        g.server = nil
+    end
+    if g.it ~= nil then
+        g.it:close()
+        g.it = nil
+    end
 end)
 
 local function assert_configured_names(server)
@@ -96,11 +88,6 @@ g.test_2_11_1 = function(g)
     --
     -- Otherwise it is not possible to connect to the instance
     -- from the test and the ready check fails on the startup.
-    --
-    -- NB: If this sequence of steps is failed in the middle (so
-    -- we don't reach g.it:close()), the g.it handle is collected
-    -- by GC and the process is killed. At worst, it is performed
-    -- on luatest termination.
     g.it = it.new()
     g.it:roundtrip("fio = require('fio')")
     g.it:roundtrip(('fio.chdir(%q)'):format(data_dir))
@@ -131,6 +118,9 @@ end
 -- config:info() if tarantool is started from a Lua script,
 -- without the declarative configuration.
 g.test_script = function(g)
+    g.server = server:new({alias = 'classic_server'})
+    g.server:start()
+
     g.server:exec(function()
         local config = require('config')
 
@@ -144,6 +134,7 @@ end
 -- Same as previous, but for tarantool started in the interactive
 -- mode, without a YAML config or a Lua script.
 g.test_interactive = function(g)
+    g.it = it.new()
     g.it:roundtrip("config = require('config')")
 
     t.assert_equals(g.it:roundtrip('config:info().hierarchy'), {})
```

I usually use this way. Don't know why I forgot about it this time.

I'm going to fix it this way.

Reply:
Done.

Reply:
Thank you. I haven't thought of this since I consider even the initialization code as the closely related to the test as different configurations can be passed. You've suggested a nice approach, hope I won't forget about it when I'll need it. A new one way to write it is also good.